### PR TITLE
Allow user to control the level of scitype checks in `machine` constructor

### LINKED ISSUE #908: Feature request: ability to convert scitype warnings into errors
Sometimes, when I do `mach = MLJ.machine(model, X, y)`, it prints one or more warnings of the form:

```
Warning: the scitype of `X` is incompatible with...
Warning: the scitype of `y` is incompatible with...
```

Is there a way to convert those warnings to errors that are thrown immediately?

### PR DESCRIPTION
**Pull request by ablaom**
**edited** (can now modify the global default value of `scitype_check_level` kwarg of `machine` constructor)

Closes https://github.com/alan-turing-institute/MLJ.jl/issues/908 .

Although not breaking, this PR has `for_a-0-point-20-release` as target, not `dev`. This is because it is not urgent and would otherwise have led to messy merge conflicts later on. 

In this PR we:

- (**enhancement**) Allow user to specify `scitype_check_level` when constructing machines, as in `machine(model, X, y, check_scitype_level=2)`, to control how strictly the constructor enforces scitype type compatibility between `model` and the data. Allow user to change the global default `scitype_check_level` using new method `scitype_check_level(i::Int)`. 

From the new docstring:

```
machine(model, args...; cache=true, scitype_check_level=1)
```
...

If `scitype_check_level > 0` (default is `1`) then the scitype of each `arg`
in `args` is computed, and this is compared with the scitypes expected
by the model, unless `args` contains `Unknown` scitypes and
`check_level < 4`, in which case no further action is taken. Warnings
are issued or errors thrown as detailed in the following table:

`check_level` | Inspect scitypes? | If `Unknown` in scitypes | If other scitype mismatch |
|:-----------:|:-----------------:|:------------------------:|:-------------------------:|
0             | ×                 |                          |                           |
1             | ✓                 |                          | warning                   |
2             | ✓                 | warning                  | warning                   |
3             | ✓                 | warning                  | error                     |
4             | ✓                 | error                    | error                     |

### CODE REVIEW

**Comment by DilumAluthge** [comment]
>  Base automatically changed from `for-a-0-point-20-release` to `dev` 9 days ago 

@ablaom Just wanted to make sure that you noticed that GitHub changed the base (target) branch of this PR.

**Comment by ablaom** [comment]
> @ablaom Just wanted to make sure that you noticed that GitHub changed the base (target) branch of this PR.

Yes, I did notice but thanks! Will resolve new conflicts after easter.

**Comment by ablaom** [comment]
@DilumAluthge I've resolved conflicts. Assuming tests pass, this is ready for your review.

**Comment by ablaom** [comment]
@DilumAluthge Do you have any time to review this PR over the next week?

File: src/MLJBase.jl [hunk 1]
```diff
@@ -297,7 +299,7 @@ export flat_values, recursive_setproperty!,
     recursive_getproperty, pretty, unwind
 
 # show.jl
-export HANDLE_GIVEN_ID, @more, @constant, @bind, color_on, color_off
+export HANDLE_GIVEN_ID, @more, @constant, color_on, color_off
```
Comment (DilumAluthge) [inline src/MLJBase.jl:302]: Why was this export removed?

Also, is that a breaking change?

File: src/machines.jl [hunk 2]
```diff
@@ -1,3 +1,34 @@
+## SCITYPE CHECK LEVEL
+
+"""
+   default_scitype_check_level()
+
+Return the current global default value for scientific type checking
+when constructing machines.
+
+   default_scitype_check_level(i)
```
Comment (DilumAluthge) [inline src/machines.jl:9]:
```suggestion
   default_scitype_check_level(i::Integer)
```

File: src/machines.jl [hunk 3]
```diff
@@ -1,3 +1,34 @@
...
+The effect of the `scitype_check_level` option in calls of the form
+`machine(model, data, scitype_check_level=...)` is summarized below:
+
+`scitype_check_level` | Inspect scitypes? | If `Unknown` in scitypes | If other scitype mismatch |
+|:-----------:|:-----------------:|:------------------------:|:-------------------------:|
+0             | ×                 |                          |                           |
+1 (startup)   | ✓                 |                          | warning                   |
+2             | ✓                 | warning                  | warning                   |
+3             | ✓                 | warning                  | error                     |
+4             | ✓                 | error                    | error                     |
```
Comment (DilumAluthge) [inline src/machines.jl:22]: It would be nice to fix the whitespace so that this table appears aligned when you look at the source code.

File: src/machines.jl [hunk 4]
```diff
@@ -161,12 +206,24 @@ mach = machine(model, X, y)
...
-to guarantee data anonymity when serializing composite models.
+Specify `cache=false` to prioritize memory management over speed.
 
 When building a learning network, `Node` objects can be substituted
-for the concrete data.
+for the concrete data but no type or dimension checks are applied.
+
+### Checks on the types of training data
+
+A model articulates it's data requirements using [scientific
```
Comment (DilumAluthge) [inline src/machines.jl:216]:
```suggestion
A model articulates its data requirements using [scientific
```

File: src/machines.jl [hunk 5]
```diff
@@ -161,12 +206,24 @@ mach = machine(model, X, y)
...
+types](https://juliaai.github.io/ScientificTypes.jl/dev/), i.e.,
+using the [`scitype`](@ref) function instead of the `typeof` function.
+
+If `scitype_check_level > 0` then the scitype of each `arg` in `args`
+is computed, and this is compared with the scitypes expected by the
+model, unless `args` contains `Unknown` scitypes and
+`scitype_check_level < 4`, in which case no further action is
+taken. Whether warnings are issued or errors thrown depends the
+level. The method `default_scitype_check_level`](@ref) controls the
+default level (`1` at startup).
```
Comment (DilumAluthge) [inline src/machines.jl:226]: After this paragraph, I think it would be helpful to insert a copy of the table from above that shows the behavior at each value of `scitype_check_level`.

Probably the easiest approach would be to define the table in a global constant string, and then interpolate that constant into each location that we want the table to appear. Then, we only have to maintain one version of the table, but we can use it multiple times.

File: src/machines.jl [hunk 6]
```diff
@@ -769,7 +832,11 @@ Returns a shallow copy of the machine to make it serializable. In particular,
 all training data is removed and, if necessary, learned parameters are replaced
 with persistent representations.
 
+<<<<<<< HEAD
+Any general purpose Julia serialization may be applied to the output of
+=======
 Any general purpose Julia serializer may be applied to the output of
+>>>>>>> dev
```
Comment (DilumAluthge) [inline src/machines.jl:839]: It looks like some Git merge conflict markers accidentally got committed.

**Comment by DilumAluthge** [comment]
I'm seeing a lot of color related changes in this PR. Are those intentional, or did something go wrong during a rebase?

**Review by DilumAluthge — APPROVED** [review]

**Comment by DilumAluthge** [comment]
> Addresses [alan-turing-institute/MLJ.jl#908](https://github.com/alan-turing-institute/MLJ.jl/issues/908) .

I would say this closes #908, right?

File: src/MLJBase.jl (hunk 1 above)
Comment (ablaom) [reply src/MLJBase.jl:302]: No @bind is some old experimental thing that was never public or documented. I think the code defining it is gone.

Not breaking.

File: src/machines.jl (hunk 5 above)
Comment (ablaom) [reply src/machines.jl:226]: Mmm. I was just thinking this doc string is already mammoth, and that's why I sucked this out. How about I change on line 225 onwards to:
 
```
level. For details, see `default_scitype_check_level`](@ref), which controls the
default level (`1` at startup).
```

File: src/machines.jl (hunk 6 above)
Comment (ablaom) [reply src/machines.jl]: fixed

File: src/machines.jl (hunk 3 above)
Comment (ablaom) [reply src/machines.jl]: done

**Comment by ablaom** [comment]
Yes.

**Comment by ablaom** [comment]
Sorry, what do you mean by "color" changes?

File: src/machines.jl (hunk 5 above)
Comment (DilumAluthge) [reply src/machines.jl:226]: Hmmm, that's a fair point. Okay, let's go with a `@ref` for now.

File: src/sources.jl [hunk 7]
```diff
@@ -129,7 +129,7 @@ function Base.show(stream::IO, object::AbstractNode)
     str = simple_repr(typeof(object))
     show_handle(object) && (str *= " $(handle(object))")
     if false
-        printstyled(IOContext(stream, :color=> SHOW_COLOR),
+        printstyled(IOContext(stream, :color=> SHOW_COLOR[]),
```
Comment (DilumAluthge) [inline src/sources.jl:132]: @ablaom Here's an example of the "color" changes that appear in this PR but seem unrelated to the PR.

Comment (ablaom) [reply src/sources.jl:132]: Oh right. No, that is intentional. It brings `SHOW_COLOR` into line with `DEFAULT_SCITYPE_CHECK_LEVEL` (introduced in this PR) and `DEFAULT_RESOURCE`.  See [here](https://github.com/JuliaAI/MLJBase.jl/blob/430eff5cc05c88e07bf16aba70b8b3b8696b6338/src/init.jl#L5). We are now now using `Ref` to avoid redefining global constants.

Comment (DilumAluthge) [reply src/sources.jl:132]: Ah, yeah a constant `Ref` sounds better than redefining globals.

**Review by DilumAluthge — APPROVED** [review]
Overall, the feature looks good!

After you have implemented your suggestion in https://github.com/JuliaAI/MLJBase.jl/pull/753#discussion_r863303906, this is good to go from my point of view.

**Comment by ablaom** [comment]
Thanks @DilumAluthge for reviewing this PR and raising the original issue.
